internal/exec/runner.go

```diff
@@ -0,0 +1,922 @@
+		return nil, err
+	}
+	if len(compileResult.Failures) > 0 {
+		return nil, newExitErrorf(255, "")
+	}
+	r.logger.Debug("protoc command exited without errors")
+	return compileResult.FileDescriptorSets, nil
+}
+
+func (r *runner) ProtocCommands(args []string, genCommands bool) error {
+	meta, err := r.getMeta(args)
+	if err != nil {
+		return err
+	}
+	r.printAffectedFiles(meta)
+	commands, err := r.newCompiler(genCommands, false).ProtocCommands(meta.ProtoSets...)
+	if err != nil {
+		return err
+	}
+	for _, command := range commands {
+		if err := r.println(command); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *runner) Lint(args []string) error {
+	meta, err := r.getMeta(args)
+	if err != nil {
+		return err
+	}
+	r.printAffectedFiles(meta)
+	if _, err := r.compile(false, false, meta); err != nil {
+		return err
+	}
+	return r.lint(meta)
+}
+
+func (r *runner) lint(meta *meta) error {
+	r.logger.Debug("calling LintRunner")
+	failures, err := r.newLintRunner().Run(meta.ProtoSets...)
+	if err != nil {
+		return err
+	}
+	if err := r.printFailures("", meta, failures...); err != nil {
+		return err
+	}
+	if len(failures) > 0 {
+		return newExitErrorf(255, "")
+	}
+	return nil
+}
+
+func (r *runner) ListLinters() error {
+	config, err := r.getConfig(r.workDirPath)
+	if err != nil {
+		return err
+	}
+	linters, err := lint.GetLinters(config.Lint)
+	if err != nil {
+		return err
+	}
+	return r.printLinters(linters)
+}
+
+func (r *runner) ListAllLinters() error {
+	return r.printLinters(lint.AllLinters)
+}
+
+func (r *runner) ListLintGroup(group string) error {
+	linters, ok := lint.GroupToLinters[strings.ToLower(group)]
+	if !ok {
+		return newExitErrorf(255, "unknown lint group: %s", strings.ToLower(group))
+	}
+	return r.printLinters(linters)
+}
+
+func (r *runner) ListAllLintGroups() error {
+	groups := make([]string, 0, len(lint.GroupToLinters))
+	for group := range lint.GroupToLinters {
+		groups = append(groups, group)
+	}
+	sort.Strings(groups)
+	for _, group := range groups {
+		if err := r.println(group); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *runner) Format(args []string, overwrite bool, diffMode bool, lintMode bool, rewrite bool) error {
+	meta, err := r.getMeta(args)
+	if err != nil {
+		return err
+	}
+	r.printAffectedFiles(meta)
+	if _, err := r.compile(false, false, meta); err != nil {
+		return err
+	}
+	return r.format(overwrite, diffMode, lintMode, rewrite, meta)
+}
+
+func (r *runner) format(overwrite bool, diffMode bool, lintMode bool, rewrite bool, meta *meta) error {
+	var retErr error
+	for _, protoSet := range meta.ProtoSets {
+		for _, protoFiles := range protoSet.DirPathToFiles {
+			for _, protoFile := range protoFiles {
+				if err := r.formatFile(overwrite, diffMode, lintMode, rewrite, meta, protoFile); err != nil {
+					if _, ok := err.(*ExitError); !ok {
+						return err
+					}
+					retErr = err
```

Comment:
I looked through the implementation, and the detail here, which is poorly abstracted, is that an `ExitError` in this context, contrary  to it's naming, is a non-system error where we had a diff on format, for example, and we don't want to exit until we process all the files. So if you pas `-l, -d`, or nothing, you want a non-zero exit code if there is a diff, but you don't want to stop until all files are processed. I refactored this a bit so it's a bit more clear, so that an error is actually an error, but it's still not great.